File: ruby/dialog/welcomedialog.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

require 'dialoghelper'

class WelcomeDialog < NSObject
  include DialogHelper
  attr_accessor :delegate, :prefix
  ib_outlet :window, :nickText, :serverCombo, :channelTable, :autoConnectCheck
  ib_outlet :okButton, :addChannelButton, :deleteChannelButton
  
  def initialize
    @prefix = 'welcomeDialog'
    @channels = []
  end
  
  def start
    NSBundle.loadNibNamed_owner('WelcomeDialog', self)
    tableViewSelectionIsChanging(nil)
    @channelTable.text_delegate = self
    ServerDialog.servers.each {|i| @serverCombo.addItemWithObjectValue(i) }
    load
    update_ok_button
    show
  end
  
  def show
    unless @window.isVisible
      @window.centerOfScreen
    end
    @window.makeKeyAndOrderFront(self)
  end
  
  def close
    @delegate = nil
    @window.close
  end
  
  def windowWillClose(sender)
    fire_event('onClose')
  end
  
  ib_action :onOk
  def onOk(sender)
    @channels.uniq!
    @channels.delete('')
    @channels.map! do |i|
      i.channelname? ? i : '#' + i
    end
    c = {
      :nick => @nickText.stringValue.to_s,
      :host => @serverCombo.stringValue.to_s,
      :channels => @channels,
      :auto_connect => @autoConnectCheck.state.to_i != 0,
    }
    fire_event('onOk', c)
    @window.close
  end
  
  ib_action :onCancel
  def onCancel(sender)
    @window.close
  end
  
  ib_action :onAddChannel
  def onAddChannel(sender)
    @channels << ''
    @channelTable.reloadData
    @channelTable.editColumn_row_withEvent_select(0, @channels.size-1, nil, true)
  end

  ib_action :onDeleteChannel
  def onDeleteChannel(sender)
    n = @channelTable.selectedRows[0]
    if n
      @channels.delete_at(n)
      @channelTable.reloadData
    end
  end
  
  def numberOfRowsInTableView(sender)
    @channels.size
  end
  
  def tableView_objectValueForTableColumn_row(sender, col, row)
    @channels[row]
  end
  
  def tableViewSelectionIsChanging(note)
    @deleteChannelButton.setEnabled(!@channelTable.selectedRows.empty?)
  end
  
  def textDidEndEditing(note)
    n = @channelTable.editedRow
    if n >= 0
      @channels[n] = note.object.textStorage.string.to_s
      @channelTable.reloadData
      @channelTable.select(n)
    end
  end
  
  def controlTextDidChange(note)
    update_ok_button
  end
  
  ib_action :onServerComboChanged
  def onServerComboChanged(sender)
    update_ok_button
  end
  
  private
  
  def load
    nick = OSX::NSUserName().gsub(/\s/, '')
    if /\A[a-z][-_a-z\d]*\z/i =~ nick
      @nickText.setStringValue(nick)
    end
  end
  
  def update_ok_button
    nick = @nickText.stringValue.to_s
    server = @serverCombo.stringValue.to_s
    @okButton.setEnabled(!nick.empty? && !server.empty?)
  end
end

File: Classes/Dialogs/WelcomeDialog.m
// LimeChat is copyrighted free software by Satoshi Nakagawa <psychs AT limechat DOT net>.
// You can redistribute it and/or modify it under the terms of the GPL version 2 (see the file GPL.txt).

#import "WelcomeDialog.h"
#import "ServerDialog.h"
#import "NSStringHelper.h"


@implementation WelcomeDialog
{
    NSMutableArray* _channels;
}

- (id)init
{
    self = [super init];
    if (self) {
        [[NSBundle mainBundle] loadNibNamed:@"WelcomeDialog" owner:self topLevelObjects:nil];

        _channels = [NSMutableArray new];

        NSArray* servers = [ServerDialog availableServers];
        for (NSString* s in servers) {
            [_hostCombo addItemWithObjectValue:s];
        }
    }
    return self;
}

- (void)dealloc
{
    _channelTable.delegate = nil;
    _channelTable.dataSource = nil;
}

- (void)show
{
    [self tableViewSelectionIsChanging:nil];
    [self updateOKButton];

    if (![self.window isVisible]) {
        [self.window center];

        NSString* username = NSUserName();
        username = [username stringByReplacingOccurrencesOfString:@" " withString:@""];

        NSRange range = [username rangeOfString:@"[a-zA-Z][-_a-zA-Z0-9]*" options:NSRegularExpressionSearch];
        if (range.location != NSNotFound) {
            _nickText.stringValue = [username substringWithRange:range];
        }
    }

    [self.window makeKeyAndOrderFront:nil];
}

- (void)close
{
    _delegate = nil;
    [self.window close];
}

- (void)onOK:(id)sender
{
    [self.window endEditingFor:nil];

    /*
     NSText* fieldEditor = [self.window fieldEditor:NO forObject:channelTable];
     if (fieldEditor) {
     [[channelTable cell] endEditing:fieldEditor];
     NSInteger n = [channelTable editedRow];
     if (n != NSNotFound) {
     NSString* s = [[[fieldEditor string] copy] autorelease];
     if (n < channels.count) {
     [channels replaceObjectAtIndex:n withObject:s];
     }
     }
     }
     */

    NSMutableSet* set = [NSMutableSet set];
    NSMutableArray* chans = [NSMutableArray array];

    for (NSString* chname in _channels) {
        NSString* s = chname;
        if (s.length > 0) {
            if (![s isChannelName]) {
                s = [@"#" stringByAppendingString:s];
            }

            if (![set containsObject:s]) {
                [chans addObject:s];
                [set addObject:s];
            }
        }
    }

    NSMutableDictionary* dic = [NSMutableDictionary dictionary];
    [dic setObject:_nickText.stringValue forKey:@"nick"];
    [dic setObject:_hostCombo.stringValue forKey:@"host"];
    [dic setObject:chans forKey:@"channels"];
    [dic setObject:@((BOOL)_autoConnectCheck.state) forKey:@"autoConnect"];

    if ([_delegate respondsToSelector:@selector(welcomeDialog:onOK:)]) {
        [_delegate welcomeDialog:self onOK:dic];
    }

    [self.window close];
}

- (void)onCancel:(id)sender
{
    [self.window close];
}

- (void)onAddChannel:(id)sender
{
    [_channels addObject:@""];
    [_channelTable reloadData];
    int row = _channels.count - 1;
    [_channelTable selectItemAtIndex:row];
    [_channelTable editColumn:0 row:row withEvent:nil select:YES];
}

- (void)onDeleteChannel:(id)sender
{
    NSInteger n = [_channelTable selectedRow];
    if (n >= 0) {
        [_channels removeObjectAtIndex:n];
        [_channelTable reloadData];
        int count = _channels.count;
        if (count <= n) n = count - 1;
        if (n >= 0) {
            [_channelTable selectItemAtIndex:n];
        }
        [self tableViewSelectionIsChanging:nil];
    }
}

- (void)controlTextDidChange:(NSNotification*)note
{
    [self updateOKButton];
}

- (void)onHostComboChanged:(id)sender
{
    [self updateOKButton];
}

- (void)updateOKButton
{
    NSString* nick = _nickText.stringValue;
    NSString* host = _hostCombo.stringValue;
    [_okButton setEnabled:nick.length > 0 && host.length > 0];
}

#pragma mark - NSTableViwe Delegate

- (void)textDidEndEditing:(NSNotification*)note
{
    NSInteger n = [_channelTable editedRow];
    if (n >= 0) {
        NSString* s = [[[[note object] textStorage] string] copy];
        [_channels replaceObjectAtIndex:n withObject:s];
        [_channelTable reloadData];
        [_channelTable selectItemAtIndex:n];
        [self tableViewSelectionIsChanging:nil];
    }
}

- (NSInteger)numberOfRowsInTableView:(NSTableView *)sender
{
    return _channels.count;
}

- (id)tableView:(NSTableView *)sender objectValueForTableColumn:(NSTableColumn *)column row:(NSInteger)row
{
    return [_channels objectAtIndex:row];
}

- (void)tableViewSelectionIsChanging:(NSNotification *)note
{
    [_deleteChannelButton setEnabled:[_channelTable selectedRow] >= 0];
}

#pragma mark - NSWindow Delegate

- (void)windowWillClose:(NSNotification*)note
{
    if ([_delegate respondsToSelector:@selector(welcomeDialogWillClose:)]) {
        [_delegate welcomeDialogWillClose:self];
    }
}

@end
